packages/@aws-cdk/aws-servicecatalog/lib/private/product-stack-asset-bucket.ts

```diff
@@ -0,0 +1,83 @@
+import * as cdk from '@aws-cdk/core';
+import { IBucket, BlockPublicAccess, Bucket } from '../../../aws-s3';
+import { BucketDeployment, ISource, Source } from '../../../aws-s3-deployment';
+import { Construct } from 'constructs';
+import { hashValues } from './util';
+
+/**
+ * A Service Catalog product stack asset bucket, which is similar in form to an Amazon S3 bucket.
+ * You can store multiple product stack assets and collectively deploy them to S3.
+ */
+export class ProductStackAssetBucket extends Construct {
+  private readonly bucketName: string;
+  private readonly bucket: IBucket;
+  private readonly assets: ISource[];
+
+  constructor(scope: Construct, id: string) {
+    super(scope, id);
+
+    this.bucketName = this.generateBucketName(id);
+
+    this.bucket = new Bucket(scope, `${id}S3Bucket`, {
+      bucketName: this.bucketName,
+      blockPublicAccess: BlockPublicAccess.BLOCK_ALL,
+    });
+
+    this.assets = [];
+  }
+
+  /**
+   * Fetch the S3 bucket.
+   *
+   * @internal
+   */
+  public getBucket(): IBucket {
+    return this.bucket;
+  }
+
+  /**
+   * Generate unique name for S3 bucket.
+   *
+   * @internal
+   */
+  private generateBucketName(id: string): string {
+    const accountId = cdk.Stack.of(this).account;
+    if (cdk.Token.isUnresolved(accountId)) {
+      throw new Error('CDK Account ID must be defined in the application environment');
+    }
+    return `product-stack-asset-bucket-${accountId}-${hashValues(id)}`;
```

Comment:
Can we use the cloudformation intrinsic `${AWS::AccountId}` instead? Do we need to know the bucketname at synth time?

Reply:
Do you mind clarifying your first question? My understanding is that buckets share a global namespace and thus, the concept of regions doesn't necessary apply.

Reply:
As for garbage collection, the great thing about having a deterministic name for the bucket is that the user will be forced to delete the buckets in the case where they destroy and try to re deploy their stacks. This way, users will avoid the scenario of having s3 buckets pile up. This being said, the only garbage collection concern is one where a user destroys their stack (with the intention of never deploying it again) and forgets to delete the asset buckets on S3.

Reply:
> Can we use the cloudformation intrinsic `${AWS::AccountId}` instead? Do we need to know the bucketname at synth time?

Yes, we are required to know the bucket name at synth time given that product stack synthesizer is required to return an expected location for the asset, and the entire appeal of product stacks is that they altogether skip deployment (they simply synthesize a template that gets uploaded to the CDK bootstrap bucket of the user). 

This being said, we tried using `${AWS::AccountId}` but the issue with doing so is that it contains a token rather than the explicit account id string. At deploy time, the token is resolved and the bucket is created. The issue comes before that, when the product stack synthesizer tries to return the expected asset location. Instead of returning product-stack-asset-bucket-1234 for the bucket name, it will return product-stack-asset-bucket-${TOKEN[Bucket.Name.1234]}

Reply:
> Do you mind clarifying your first question? My understanding is that buckets share a global namespace and thus, the concept of regions doesn't necessary apply.

A global namespace does not mean that it is a global service. If I publish the assets to a bucket in `us-east-1`, but I consume the assets in a stack deployment in `us-east-2` I've created a cross region dependency. What if `us-east-1` is experiencing an outage and I need to deploy my failover stack in `us-east-2`? I won't be able to deploy because I have a dependency on `us-east-1`.

> As for garbage collection...

Maybe I misunderstand how this feature works. I'm more talking about garbage collection for the assets themselves. When I change the lambda code and deploy, does the new asset go to the same bucket? Overtime  this might lead to a lot of outdated assets.